src/source-dpdk.c

```diff
@@ -0,0 +1,494 @@
+
+    if (initdata == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "DPDK Configuration failed");
+        dpdk_config->DerefFunc(dpdk_config);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    DPDKThreadVars *ptv = SCCalloc(1, sizeof(DPDKThreadVars));
+    if (unlikely(ptv == NULL)) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Unable to allocate memory");
+        dpdk_config->DerefFunc(dpdk_config);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    ptv->tv = tv;
+    ptv->pkts = 0;
+    ptv->bytes = 0;
+    ptv->livedev = LiveGetDevice(dpdk_config->iface);
+
+#ifdef PACKET_STATISTICS
+    ptv->capture_dpdk_packets = StatsRegisterCounter("capture.packets", ptv->tv);
+    ptv->capture_dpdk_drops = StatsRegisterCounter("capture.drops", ptv->tv);
+#endif
+
+    ptv->copy_mode = dpdk_config->copy_mode;
+    ptv->checksum_mode = dpdk_config->checksum_mode;
+
+    ptv->threads = dpdk_config->threads;
+    ptv->port_id = dpdk_config->port_id;
+    ptv->out_port_id = dpdk_config->out_port_id;
+    uint16_t queue_id = SC_ATOMIC_ADD(dpdk_config->queue_id, 1);
+    ptv->queue_id = queue_id;
+    // pass the pointer to the mempool and then forget about it. Mempool is freed in thread deinit.
+    ptv->pkt_mempool = dpdk_config->pkt_mempool_array[queue_id];
+    dpdk_config->pkt_mempool_array[queue_id] = NULL;
+
+    // the last thread starts the device
+    if (queue_id == dpdk_config->threads - 1) {
+        retval = rte_eth_dev_start(ptv->port_id);
+        if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) during device startup of port %u", retval,
+                    ptv->port_id);
+            dpdk_config->DerefFunc(dpdk_config);
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+
+        retval = dpdk_config->flags & DPDK_PROMISC ? rte_eth_promiscuous_enable(ptv->port_id)
+                                                   : rte_eth_promiscuous_disable(ptv->port_id);
+        if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) when enabling promiscuous mode on port %u",
+                    retval, ptv->port_id);
+            dpdk_config->DerefFunc(dpdk_config);
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+    }
+
+    if (rte_eth_dev_socket_id(ptv->port_id) != (int)rte_socket_id()) {
+        SCLogWarning(SC_WARN_DPDK_CONF,
+                "NIC on a different NUMA node as lcore %u is initialized on", rte_socket_id());
+    }
+
+    *data = (void *)ptv;
+    dpdk_config->DerefFunc(dpdk_config);
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/**
+ * \brief This function prints stats to the screen at exit.
+ * \param tv pointer to ThreadVars
+ * \param data pointer that gets cast into DPDKThreadVars for ptv
+ */
+void ReceiveDPDKThreadExitStats(ThreadVars *tv, void *data)
+{
+    SCEnter();
+    DPDKThreadVars *ptv = (DPDKThreadVars *)data;
+
+#ifdef PACKET_STATISTICS
+    DPDKDumpCounters(ptv);
+    SCLogPerf("(%s) DPDK: Packets %" PRIu64 " dropped: %" PRIu64 "", tv->name,
```

Comment:
highly recommend to add packet drops and invalid counters too

Reply:
what about missed opprotunites for rx_burst and no rx-mbuf. you are already pulling DPDK stats.

Reply:
The counters for no rx-mbuf are already included.